src/effects/effectslot.cpp

```diff
@@ -268,3 +272,62 @@ void EffectSlot::slotEffectMetaParameter(double v, bool force) {
         pParameterSlot->onEffectMetaParameterChanged(v, force);
     }
 }
+
+QDomElement EffectSlot::toXml(QDomDocument* doc) const {
+    QDomElement effectElement = doc->createElement(EffectXml::Effect);
+    if (m_pEffect == nullptr) {
+        return effectElement;
+    }
+
+    QDomElement metaKnobElement = doc->createElement(EffectXml::EffectMetaParameter);
+    XmlParse::addElement(*doc, effectElement,
+                         EffectXml::EffectMetaParameter,
+                         QString::number(m_pControlMetaParameter->get()));
+    EffectManifest manifest = m_pEffect->getManifest();
+    XmlParse::addElement(*doc, effectElement,
+                         EffectXml::EffectId, manifest.id());
+    XmlParse::addElement(*doc, effectElement,
+                         EffectXml::EffectVersion, manifest.version());
+
+    QDomElement parametersElement = doc->createElement(EffectXml::ParametersRoot);
+
+    for (const auto& pParameter : m_parametersById) {
+        QDomElement parameterElement = pParameter->toXml(doc);
+        XmlParse::addElement(*doc, parameterElement,
+                             EffectXml::ParameterId,
+                             pParameter->getManifest().id());
+        parametersElement.appendChild(parameterElement);
+    }
+
+    effectElement.appendChild(parametersElement);
+
+    return effectElement;
+}
+
+void EffectSlot::loadEffectSlotFromXml(const QDomElement& effectElement) {
+    if (effectElement.text().isEmpty()) {
+        return;
+    }
+
+    m_pControlMetaParameter->set(XmlParse::selectNodeDouble(effectElement,
+                                                            EffectXml::EffectMetaParameter));
+
+    QDomElement parametersElement = XmlParse::selectElement(effectElement,
+                                                            EffectXml::ParametersRoot);
+    if (parametersElement.text().isEmpty()) {
+        return;
+    }
+
+    QDomNodeList parametersNodeList = parametersElement.childNodes();
+    for (int i = 0; i < parametersNodeList.size(); ++i) {
+        QDomNode parameterNode = parametersNodeList.at(i);
+        if (parameterNode.isElement()) {
+            const QString id = XmlParse::selectNodeQString(parameterNode,
+                                                           EffectXml::ParameterId);
+            if (m_parametersById.contains(id)) {
```

Comment:
This pattern does two O(log n) lookups in the QMap, you can avoid it like this:

```c++
EffectParameterSlotBasePointer pParameterSlot = m_parametersById.value(id);
if (pParameterSlot) { 
}
```
since QMap::value returns a default-constructed value if it's not present